Add Support for On-Demand Capacity Reservations (ODCRs) in Cluster Configuration File

hanwen-pcluste opened the pull request:
### Description of changes
Add ODCR configuration under queue and compute resource levels: See commits descriptions for details

**To Do**
In a separate PR, we will create a dedicated integration test to verify target ODCR

### Tests
A unit test is created for the new validator. Integration test (test_efa) is changed to use ODCR from configuration file. Manually verified the [run_instances override](https://docs.aws.amazon.com/parallelcluster/latest/ug/launch-instances-odcr-v3.html) has higher priority than the parameters in cluster config

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [ ] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

hanwen-cluster committed:
Support On-Demand Capacity Reservations in Cluster Configuration File

1. The section can appear both on queue level and compute resource level. The section on compute resource level takes precedence over the queue level.

2. The capacity reservation specification is passed into the launch templates used by compute nodes. This approach is forward compatible if ParallelCluster uses EC2 fleet

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Integ-tests: Change test_efa to use ODCR from cluster configuration file

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Add policy to allow Head node to use ODCR

This commit also adds a validation to prevent the two parameters coexist in ODCR section

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Add validation to make sure the ODCR has the right instance type and availability zone

Signed-off-by: Hanwen <hanwenli@amazon.com>

chenwany commented on tests/integration-tests/tests/efa/test_efa.py:
```diff
@@ -54,19 +53,8 @@ def test_efa(
-        head_node_instance=head_node_instance,
-        bucket_name=bucket_name,
-    )
+    cluster_config = pcluster_config_reader(max_queue_size=max_queue_size, head_node_instance=head_node_instance)
```
s3_bucket_factory can be removed from line 38 since we don't use it any more

hanwen-pcluste replied:
Done

chenwany commented on cli/tests/pcluster/validators/test_ec2_validators.py:
```diff
@@ -457,3 +458,69 @@ def test_placement_group_validator(
+        ),
+    ],
+)
+def test_capacity_reservation_validator_validator(
```
Can we also add the test for CapacityReservationResourceGroupValidator?

hanwen-pcluste replied:
Done

chenwany commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -2401,6 +2418,22 @@ def _register_validators(self):
                         instance_type=compute_resource.instance_type,
                         instance_type_data=instance_types_data[compute_resource.instance_type],
                     )
+                # The validation below has to be in cluster config class instead of queue class
```
Can line 2382 and these two validators to be added under CommonSchedulerClusterConfig? The reason is that from the schema, it seems that it also support for scheduler plugin schema

hanwen-pcluste replied:
Which validators? Sorry, I changed the line number with some rebase

chenwany replied:
CapacityReservationValidator and CapacityReservationResourceGroupValidator.
I think the field `capacity_reservation_target` is both for slurm and scheduler plugin in the schema. 
Why are the validators added for only class SlurmClusterConfig instaed of its parent class?
`class SlurmClusterConfig(CommonSchedulerClusterConfig)`

hanwen-pcluste replied:
After discussion with chenwany, ODCR should be supported by both Slurm and Scheduler Plugin. I made relevant changes. Thank you!

dreambeyondorange commented on cli/src/pcluster/aws/resource_groups.py:
```diff
@@ -0,0 +1,31 @@
+        resources = self._client.list_group_resources(Group=group)["Resources"]
+        for resource in resources:
+            if resource["Identifier"]["ResourceType"] == "AWS::EC2::CapacityReservation":
+                capacity_reservation_ids.append(re.match("(.*)(cr-.*)", resource["Identifier"]["ResourceArn"]).group(2))
```
is there a way to match exactly so there is only one result?

dreambeyondorange replied:
It might also be good to clarify what the matches are using the name syntax `?P<match_name>`
```
>>> m = re.match(r"(?P<first_name>\w+) (?P<last_name>\w+)", "Malcolm Reynolds")
>>> m.group('first_name')
'Malcolm'
>>> m.group('last_name')
'Reynolds'
```

hanwen-pcluste replied:
> is there a way to match exactly so there is only one result?

I made the regex stricter, but not sure if there is a exact match

The second comment is addressed

hanwen-cluster committed:
Add policy to allow ParallelCluster API to list group resources

This policy is required to run validators for `CapacityReservationResourceGroupArn`.

The policy of describing capacity reservation is added in this commit. Because it is covered by existing policy: `ec2:Describe*`

The same policy is added to integration tests

Signed-off-by: Hanwen <hanwenli@amazon.com>

chenwany commented on cli/src/pcluster/templates/cluster_stack.py:
```diff
@@ -1459,6 +1459,17 @@ def _add_compute_resource_launch_template(
             conditional_template_properties.update({"instance_type": compute_resource.instance_type})
 
+        capacity_reservation_specification = None
+        if isinstance(queue, (SlurmQueue, SchedulerPluginQueue)):
```
Is this check `f isinstance(queue, (SlurmQueue, SchedulerPluginQueue))` necessary? 
`def _add_compute_resource_launch_template` function is under class `ComputeFleetConstruct`, ComputeFleetConstruct is not for awsbatch scheduler.
See code:
```
        if not self._condition_is_batch():
            self.compute_fleet_resources = ComputeFleetConstruct(
```

hanwen-pcluste replied:
Done

chenwany commented on cli/src/pcluster/templates/cdk_builder_utils.py:
```diff
@@ -654,6 +654,32 @@ def _build_policy(self) -> List[iam.PolicyStatement]:
                             ]
                         )
 
+        if self._config.scheduling.scheduler == "slurm":
```
`if self._config.scheduling.scheduler != "awsbatch"`?

hanwen-pcluste replied:
Done

hanwen-pcluste committed:
Merge branch 'develop' into release-3.2origin
